torchsynth/module.py

```diff
@@ -632,4 +646,62 @@ def forward(self) -> Tuple[T, T]:
         # But usually parameter stuff is not the bottleneck
         params = torch.stack([p.data for p in self.torchparameters.values()])
         params = torch.pow(params, exponent=self.exponent)
+
+        # This returns a tensor with shape (n_modes, batch_size), should it
+        # (batch_size, n_modes) to be more aligned with the rest of the code base?
         return params / torch.sum(params, dim=0)
+
+
+class NormalizedFaderBank(SynthModule):
+    """
+    A bank of faders with values between 0 and 1. Each fader parameter has a
+    curve which puts more emphasis on lower values or higher values. The output
+    of the faders is normalized to sum to 1.0 if the sum of the faders is greater
+    than 1.0 prior to normalization.
+    """
+
+    def __init__(
+        self,
+        synthglobals: SynthGlobals,
+        curves: List[float],
+        **kwargs: Dict[str, T],
+    ):
+        # Need to create the parameter ranges before calling super().__init
+        self.parameter_ranges = [
+            ModuleParameterRange(
+                0.0,
+                1.0,
+                curve=curves[i],
+                name=f"fader{i}",
+                description=f"fader {i} value",
+            )
+            for i in range(len(curves))
+        ]
+        super().__init__(synthglobals, **kwargs)
+
+    def forward(self) -> T:
+        """
+        Return all the fader values in a tensor with shape (batch_size, n_faders)
+        """
+        params = torch.stack([self.p(p) for p in self.torchparameters])
+
+        # If the sum of faders is greater then 1.0 then normalize
+        summed = torch.sum(params, dim=0)
+        return torch.where(summed > 1.0, params / summed, params).T
```

Comment:
Is this the right way to think about this? I want to allow for the sum of the parameters to be between 0 and 1, but never greater than 1 while also preserving the relationship between the parameters. Not sure if this is going to lead to weird behavior though. Is this confusing?

Reply:
idk? why not normalize all of them?

Reply:
So we want to be able to have the sum of the parameters be any where between 0 and 1. If we normalize them all then the sum will always be 1. It's like having a mixing board that allows you to have a mixture level up to 1, and then when you exceed that it automatically balances the mixture to preserve the ratio between the signals without going above 1. @maxsolomonhenry any thoughts about this?

Reply:
It would be consistent with that we're doing in the noise module, for example. Maybe not the most intuitive for usability, but lets us make the most of the [-1, 1] sample range. 

The rub with either is that we lose any meaningful variability in volume. But since we're batch normalizing in a lot of the eventual learning, I dunno it's probably fine.